apps/astarte_data_updater_plant/test/astarte_data_updater_plant/device_disconnection_test.exs

```diff
@@ -0,0 +1,113 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.DataUpdaterPlant.DeviceDisconnectionTest do
+  use ExUnit.Case, async: true
+  import Mox
+
+  alias Astarte.DataUpdaterPlant.DatabaseTestHelper
+  alias Astarte.Core.Device
+  alias Astarte.DataUpdaterPlant.DataUpdater
+  alias Astarte.DataAccess.Realms.Realm
+  alias Astarte.DataAccess.Repo
+  alias Astarte.DataAccess.Devices.Device, as: DeviceSchema
+  alias Astarte.DataUpdaterPlant.AMQPTestHelper
+
+  import Ecto.Query
+
+  setup :verify_on_exit!
+
+  setup do
+    realm = "autotestrealm#{System.unique_integer([:positive])}"
+    {:ok, _keyspace_name} = DatabaseTestHelper.create_test_keyspace(realm)
+
+    on_exit(fn ->
+      DatabaseTestHelper.destroy_local_test_keyspace(realm)
+    end)
+
+    {:ok, _pid} = AMQPTestHelper.start_link()
+    %{realm: realm}
+  end
+
+  test "device disconnection test", %{realm: realm} do
+    AMQPTestHelper.clean_queue()
+
+    keyspace_name = Realm.keyspace_name(realm)
+    encoded_device_id = "f0VMRgIBAQAAAAAAAAAAAA"
+    {:ok, device_id} = Device.decode_device_id(encoded_device_id)
+
+    received_msgs = 45000
+    received_bytes = 4_500_000
+    existing_introspection_map = %{"com.test.LCDMonitor" => 1, "com.test.SimpleStreamTest" => 1}
+
+    insert_opts = [
+      introspection: existing_introspection_map,
+      total_received_msgs: received_msgs,
+      total_received_bytes: received_bytes,
+      groups: ["group1"]
+    ]
+
+    DatabaseTestHelper.insert_device(realm, device_id, insert_opts)
+    ### -----------------------------------------------------------####################
+
+    DataUpdater.handle_disconnection(
+      realm,
+      encoded_device_id,
+      gen_tracking_id(),
+      make_timestamp("2017-10-09T14:30:45+00:00")
+    )
+
+    DataUpdater.dump_state(realm, encoded_device_id)
+
+    device_query =
+      from d in DeviceSchema,
+        prefix: ^keyspace_name,
+        where: d.device_id == ^device_id,
+        select: %{
+          connected: d.connected,
+          total_received_msgs: d.total_received_msgs,
+          total_received_bytes: d.total_received_bytes,
+          exchanged_msgs_by_interface: d.exchanged_msgs_by_interface,
+          exchanged_bytes_by_interface: d.exchanged_bytes_by_interface
+        }
+
+    device_row = Repo.one(device_query)
+
+    assert device_row == %{
+             connected: false,
+             total_received_msgs: 45000,
+             total_received_bytes: 4_500_000,
```

Comment:
Nit: since we know beforehand what values we want here, you can just do
```suggestion
             total_received_msgs: received_msgs,
             total_received_bytes: received_bytes,
```